File: erpc_python/erpc/basic_codec.py
#!/usr/bin/env python

# Copyright (c) 2015 Freescale Semiconductor, Inc.
# Copyright 2016-2017 NXP
# All rights reserved.
#
# SPDX-License-Identifier: BSD-3-Clause

import struct
from .codec import (MessageType, MessageInfo, Codec, CodecError)


class BasicCodec(Codec):
    """Version of this codec.

    Args:
        Codec (code.Codec): Inherit and implement codec interface functions.
    """

    BASIC_CODEC_VERSION = 1

    def start_write_message(self, msgInfo):
        header = (self.BASIC_CODEC_VERSION << 24) \
            | ((msgInfo.service & 0xff) << 16) \
            | ((msgInfo.request & 0xff) << 8) \
            | (msgInfo.type.value & 0xff)
        self.write_uint32(header)
        self.write_uint32(msgInfo.sequence)

    def _write(self, fmt, value):
        self._buffer += struct.pack(fmt, value)
        self._cursor += struct.calcsize(fmt)

    def write_bool(self, value):
        self._write('<?', value)

    def write_int8(self, value):
        self._write('<b', value)

    def write_int16(self, value):
        self._write('<h', value)

    def write_int32(self, value):
        self._write('<i', value)

    def write_int64(self, value):
        self._write('<q', value)

    def write_uint8(self, value):
        self._write('<B', value)

    def write_uint16(self, value):
        self._write('<H', value)

    def write_uint32(self, value):
        self._write('<I', value)

    def write_uint64(self, value):
        self._write('<Q', value)

    def write_float(self, value):
        self._write('<f', value)

    def write_double(self, value):
        self._write('<d', value)

    def write_string(self, value):
        self.write_binary(value.encode())

    def write_binary(self, value):
        self.write_uint32(len(value))
        self._buffer += value

    def start_write_list(self, length):
        self.write_uint32(length)

    def start_write_union(self, discriminator):
        self.write_uint32(discriminator)

    def write_null_flag(self, flag):
        self.write_uint8(1 if flag else 0)

    def start_read_message(self):
        """ Returns 4-tuple of msgType, service, request, sequence.

        Raises:
            CodecError: Raise this error when unsupported codec version doesn't match.

        Returns:
            codec.MessageInfo: 4-tuple of msgType, service, request, sequence.
        """
        header = self.read_uint32()
        sequence = self.read_uint32()
        version = header >> 24
        if version != self.BASIC_CODEC_VERSION:
            raise CodecError("unsupported codec version %d" % version)
        service = (header >> 16) & 0xff
        request = (header >> 8) & 0xff
        msgType = MessageType(header & 0xff)
        return MessageInfo(type=msgType, service=service, request=request, sequence=sequence)

    def _read(self, fmt):
        result = struct.unpack_from(fmt, self._buffer, self._cursor)
        self._cursor += struct.calcsize(fmt)
        return result[0]

    def read_bool(self):
        return self._read('<?')

    def read_int8(self):
        return self._read('<b')

    def read_int16(self):
        return self._read('<h')

    def read_int32(self):
        return self._read('<i')

    def read_int64(self):
        return self._read('<q')

    def read_uint8(self):
        return self._read('<B')

    def read_uint16(self):
        return self._read('<H')

    def read_uint32(self):
        return self._read('<I')

    def read_uint64(self):
        return self._read('<Q')

    def read_float(self):
        return self._read('<f')

    def read_double(self):
        return self._read('<d')

    def read_string(self):
        return self.read_binary().decode()

    def read_binary(self):
        length = self.read_uint32()
        data = self._buffer[self._cursor:self._cursor+length]
        self._cursor += length
        return data

    def start_read_list(self):
        """ Function which should be called on list de-serialization.

        Returns:
            int: Int of list length.
        """
        return self.read_uint32()

    def start_read_union(self):
        """ Function which should be called on union de-serialization.

        Returns:
            int: Int of union discriminator.
        """
        return self.read_int32()

    def read_null_flag(self):
        return self.read_uint8()

File: erpc_java/src/main/java/io/github/embeddedrpc/erpc/codec/BasicCodec.java
/*
 * Copyright 2023 NXP
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

package  io.github.embeddedrpc.erpc.codec;

import  io.github.embeddedrpc.erpc.auxiliary.MessageInfo;
import  io.github.embeddedrpc.erpc.auxiliary.MessageType;
import  io.github.embeddedrpc.erpc.auxiliary.Utils;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Basic implementation of the Codec.
 */
@SuppressWarnings("checkstyle:MagicNumber")
public final class BasicCodec implements Codec {
    private static final int BASIC_CODEC_VERSION = 1;
    private static final int DEFAULT_BUFFER_SIZE = 256;

    private ByteBuffer buffer;

    /**
     * Basic codec constructor. Create empty buffer.
     */
    public BasicCodec() {
        this.reset();
    }

    /**
     * Basic codec constructor. Create buffer from byte array.
     *
     * @param byteArray byte array
     */
    public BasicCodec(final byte[] byteArray) {
        this.buffer = ByteBuffer.wrap(byteArray).order(ByteOrder.LITTLE_ENDIAN);
    }

    private void prepareForWrite(final int bytesToWrite) {
        if (this.buffer.remaining() < bytesToWrite) {
            ByteBuffer newBuffer = ByteBuffer.allocate(
                    (this.buffer.capacity() + bytesToWrite) * 2
            );
            newBuffer.order(ByteOrder.LITTLE_ENDIAN);
            this.buffer.flip();

            newBuffer.put(this.buffer);
            this.buffer = newBuffer;
        }
    }

    @Override
    public void reset() {
        this.buffer = ByteBuffer.allocate(DEFAULT_BUFFER_SIZE)
                .order(ByteOrder.LITTLE_ENDIAN);
    }

    @Override
    public byte[] array() {
        int position = buffer.position();
        byte[] data = new byte[position];
        buffer.position(0);
        buffer.get(data);
        buffer.position(position);
        return data;
    }

    @Override
    public void setArray(final byte[] array) {
        this.buffer = ByteBuffer.wrap(array).order(ByteOrder.LITTLE_ENDIAN);
    }

    @Override
    public void startWriteMessage(final MessageInfo msgInfo) {
        long header = (BASIC_CODEC_VERSION << 24)
                | ((msgInfo.service() & 0xff) << 16)
                | ((msgInfo.request() & 0xff) << 8)
                | (msgInfo.type().getValue() & 0xff);

        this.writeUInt32(header);
        this.writeUInt32(msgInfo.sequence());
    }

    @Override
    public void writeBool(final Boolean value) {
        prepareForWrite(1);
        this.buffer.put(value ? (byte) 1 : (byte) 0);
    }

    @Override
    public void writeInt8(final byte value) {
        prepareForWrite(1);
        this.buffer.put(value);
    }

    @Override
    public void writeInt16(final short value) {
        prepareForWrite(2);
        this.buffer.putShort(value);
    }

    @Override
    public void writeInt32(final int value) {
        prepareForWrite(4);
        this.buffer.putInt(value);
    }

    @Override
    public void writeInt64(final long value) {
        prepareForWrite(8);
        this.buffer.putLong(value);
    }

    @Override
    public void writeUInt8(final short value) {
        Utils.checkUInt8(value);
        prepareForWrite(1);
        this.buffer.put(Utils.uInt8toByte(value));
    }

    @Override
    public void writeUInt16(final int value) {
        Utils.checkUInt16(value);
        prepareForWrite(2);
        this.buffer.putShort(Utils.uInt16toShort(value));
    }

    @Override
    public void writeUInt32(final long value) {
        Utils.checkUInt32(value);
        prepareForWrite(4);
        this.buffer.putInt(Utils.uInt32toInt(value));
    }

    @Override
    public void writeUInt64(final long value) {
        throw new UnsupportedOperationException(
                "Java implementation of the eRPC does not support 'uint64'"
        );
    }

    @Override
    public void writeFloat(final float value) {
        prepareForWrite(4);
        this.buffer.putFloat(value);
    }

    @Override
    public void writeDouble(final double value) {
        prepareForWrite(8);
        this.buffer.putDouble(value);
    }

    @Override
    public void writeString(final String value) {
        this.writeBinary(value.getBytes());
    }

    @Override
    public void writeBinary(final byte[] value) {
        this.writeInt32(value.length);
        prepareForWrite(value.length);
        this.buffer.put(value);
    }

    @Override
    public void startWriteList(final int length) {
        this.writeUInt32(length);
    }

    @Override
    public void startWriteUnion(final int discriminator) {
        this.writeUInt32(discriminator);
    }

    @Override
    public void writeNullFlag(final int value) {
        this.writeUInt32(value != 0 ? 1 : 0);
    }

    @Override
    public MessageInfo startReadMessage() {
        int header = (int) this.readUInt32();
        int sequence = (int) this.readUInt32();
        int version = header >> 24;

        if (version != BASIC_CODEC_VERSION) {
            throw new CodecError("Unsupported codec version" + version);
        }

        int service = (header >> 16) & 0xff;
        int request = (header >> 8) & 0xff;
        MessageType msgType = MessageType.getMessageType(header & 0xff);

        return new MessageInfo(msgType, service, request, sequence);
    }

    @Override
    public Boolean readBool() {
        return this.buffer.get() != 0;
    }

    @Override
    public byte readInt8() {
        return this.buffer.get();
    }

    @Override
    public short readInt16() {
        return this.buffer.getShort();
    }

    @Override
    public int readInt32() {
        return this.buffer.getInt();
    }

    @Override
    public long readInt64() {
        return this.buffer.getLong();
    }

    @Override
    public short readUInt8() {
        return Utils.byteToUInt8(this.buffer.get());
    }

    @Override
    public int readUInt16() {
        return Utils.shortToUInt16(this.buffer.getShort());
    }

    @Override
    public long readUInt32() {
        return Utils.intToUInt32(this.buffer.getInt());
    }

    @Override
    public long readUInt64() {
        throw new UnsupportedOperationException(
                "Java implementation of the eRPC does not support 'uint64'"
        );
    }

    @Override
    public float readFloat() {
        return this.buffer.getFloat();
    }

    @Override
    public double readDouble() {
        return this.buffer.getDouble();
    }

    @Override
    public String readString() {
        return new String(this.readBinary());
    }

    @Override
    public byte[] readBinary() {
        long length = readUInt32();
        byte[] data = new byte[(int) length];
        this.buffer.get(data, 0, (int) length);

        return data;
    }

    @Override
    public long startReadList() {
        return this.readUInt32();
    }

    @Override
    public int startReadUnion() {
        return this.readInt8();
    }

    @Override
    public Boolean readNullFlag() {
        return this.readUInt8() != 0;
    }

    @Override
    public ByteBuffer getBuffer() {
        return buffer;
    }
}
